File: java-automation/src/test/java/com/myecommerce/automation/driver/ports/BrowserPort.java
package com.myecommerce.automation.driver.ports;

import java.util.List;
import java.util.Map;
import java.util.function.Supplier;

public interface BrowserPort {

    // Navigation
    void navigateTo(String url);
    void navigateBack();
    String currentUrl();

    // Element state
    boolean isVisible(String css);
    boolean isPresent(String css);
    boolean isEnabled(String css);
    boolean isSelected(String css);
    int count(String css);

    // Element content
    String text(String css);
    String attribute(String css, String attr);

    // Nth-element queries (flat list, zero-based index)
    boolean isNthEnabled(String css, int index);
    boolean isNthSelected(String css, int index);
    String nthAttribute(String css, int index, String attr);
    String nthText(String css, int index);

    // Scoped queries — within the nth element matching parentCss
    boolean isSelectedWithin(String parentCss, int parentIndex, String childCss);
    String attributeWithin(String parentCss, int parentIndex, String childCss, String attr);

    // Actions
    void click(String css);
    void clickNth(String css, int index);
    void clickXpath(String xpath);
    void sendKeys(String css, String text, boolean submitAfter);
    void setReactInputValue(String css, String value);

    // JavaScript extraction
    List<Map<String, String>> extractAllViaScript(String script);
    Object executeScript(String script, Object... args);

    // Waits
    void waitUntilVisible(String css);
    void waitUntilPresent(String css);
    void waitUntilCountMoreThan(String css, int count);
    void waitUntilUrlContains(String fragment);
    void waitUntilUrlMatches(String regex);
    void waitUntilAttributeChanges(String css, int index, String attr, String previousValue);
    void waitUntilAnyPresent(String... cssList);
    void waitUntilCondition(Supplier<Boolean> condition, int timeoutSeconds);
}

File: kotlin-automation/src/test/kotlin/com/myecommerce/automation/driver/ports/BrowserPort.kt
package com.myecommerce.automation.driver.ports

interface BrowserPort {

    // Navigation
    fun navigateTo(url: String)
    fun navigateBack()
    fun currentUrl(): String

    // Element state
    fun isVisible(css: String): Boolean
    fun isPresent(css: String): Boolean
    fun isEnabled(css: String): Boolean
    fun isSelected(css: String): Boolean
    fun count(css: String): Int

    // Element content
    fun text(css: String): String
    fun attribute(css: String, attr: String): String

    // Nth-element queries (flat list, zero-based index)
    fun isNthEnabled(css: String, index: Int): Boolean
    fun isNthSelected(css: String, index: Int): Boolean
    fun nthAttribute(css: String, index: Int, attr: String): String
    fun nthText(css: String, index: Int): String

    // Scoped queries — within the nth element matching parentCss
    fun isSelectedWithin(parentCss: String, parentIndex: Int, childCss: String): Boolean
    fun attributeWithin(parentCss: String, parentIndex: Int, childCss: String, attr: String): String

    // Actions
    fun click(css: String)
    fun clickNth(css: String, index: Int)
    fun clickXpath(xpath: String)
    fun sendKeys(css: String, text: String, submitAfter: Boolean)
    fun setReactInputValue(css: String, value: String)

    // JavaScript extraction
    fun extractAllViaScript(script: String): List<Map<String, String>>
    fun executeScript(script: String, vararg args: Any?): Any?

    // Waits
    fun waitUntilVisible(css: String)
    fun waitUntilPresent(css: String)
    fun waitUntilCountMoreThan(css: String, count: Int)
    fun waitUntilUrlContains(fragment: String)
    fun waitUntilUrlMatches(regex: String)
    fun waitUntilAttributeChanges(css: String, index: Int, attr: String, previousValue: String)
    fun waitUntilAnyPresent(vararg cssList: String)
    fun waitUntilCondition(condition: () -> Boolean, timeoutSeconds: Int)
    fun tryWaitUntilPresent(css: String, timeoutSeconds: Int): Boolean
}
